metacat-common/src/main/java/com/netflix/metacat/common/MetacatRequestContext.java

```diff
@@ -65,18 +87,141 @@
      * @param clientId        client id
      * @param jobId           job id
      * @param dataTypeContext data type context
+     * @param apiUri          the uri of rest api
+     * @param scheme          http, thrift, internal, etc.
      */
-    public MetacatRequestContext(
+    protected MetacatRequestContext(
         @Nullable final String userName,
         @Nullable final String clientAppName,
         @Nullable final String clientId,
         @Nullable final String jobId,
-        @Nullable final String dataTypeContext
+        @Nullable final String dataTypeContext,
+        final String apiUri,
+        final String scheme
     ) {
         this.userName = userName;
         this.clientAppName = clientAppName;
         this.clientId = clientId;
         this.jobId = jobId;
         this.dataTypeContext = dataTypeContext;
+        this.apiUri = apiUri;
+        this.scheme = scheme;
+    }
+
+    /**
+     * builder class for MetacatRequestContext.
+     * @return the builder class for MetacatRequestContext
+     */
+    public static MetacatRequestContext.MetacatRequestContextBuilder builder() {
+        return new MetacatRequestContext.MetacatRequestContextBuilder();
+    }
+
+    /**
+     * MetacatRequestContext builder class.
+     */
+    public static class MetacatRequestContextBuilder {
```

Comment:
I'm pretty sure there's a bug in here. Since the `apiUri` and the `scheme` can't be null in the actual instance this builder has no required parameters or default settings for those variables. If the setters aren't called the fields in this builder remain `null` and when the `build` method is called nulls will be passed to the parameterized constructor above. In the default constructor below you should probably assign your api and scheme to whatever default values you wanted.